KEP-1731 : Publishing Kubernetes packages on community infrastructure (#3434)
kubernetes/enhancements: keps/sig-release/1731-publishing-packages/README.md

>   - [Non-Goals](#non-goals)

Author: This is great feedback, @saschagrunert TYSM

>   - [Non-Goals](#non-goals)

Author: @saschagrunert can you please outline what you would like to see happen here regarding the re-scoping of this KEP?Specifically, do you want this KEP to super-cede and replace #2503 and should we record that here?

>   - [Non-Goals](#non-goals)

Reviewer: The main re-scope would be that we as SIG Release own the process of automatically building and publishing the packages. I think this way we implicitly achieve the main goal of #2503.

What are your thoughts on that @Klaven, @shekhar-rajak, @ncdc?

- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

## Summary

This document describes how deb and rpm packages get published in the same way as the currently [documented package mirror] as part of cutting a release, with the tooling the release management team has at hand ([krel] and others from [k/release], [k/k/build] and potentially other places).

[documented package mirror]: https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/#install-using-native-package-management
[gcbmgr]: https://github.com/kubernetes/release/tree/master/gdcbmgr
[k/release]: https://github.com/kubernetes/release
[k/k/build]: https://github.com/kubernetes/kubernetes/tree/master/build

## Motivation

Currently:

- Only Googlers can publish packages.
- The packages get published on Google infrastructure.

> - After publishing a new release but before sending out the release

Reviewer 2: Nit, there's a typo/grammar mistake:

> ... branch management team can continue and send notifications ~can be sent out~.

Reviewer: Fixed

- After publishing a new release but before sending out the release notification
  emails the process needs to be paused. Googlers need to build and publish the
  deb and rpm packages before the branch manager can continue and send the
  release announcement.
- We can only publish packages for stable releases right now.
- We use different packages in CI then we officially release.

This all prolongs the release process, it is a hard dependency on a small group of people from one company (and its infrastructure), and we only ever publish and test packages very late in the release process.

### Goals

The whole process should be folded into the release tooling, it should be part of the release process, and should not involve anyone other than the release managers.
For each release the release team cuts, packages should also be generated and published automatically.

There should be multiple channels users can subscribe to: **stable**, **dev**, and **nightly**.

### Non-Goals

The original version of this KEP had the following statement:

> The actual package generation is a different problem that is discussed in [this KEP][pkg-gen-kep].

> Based on recent review of this KEP @saschagrunert wants SIG Release to own

Reviewer 3: ```suggestion
Based on a review of this KEP, Kubernetes contributor @saschagrunert wants SIG Release to own
```
“recent” would go stale

Reviewer: Thanks, fixed as suggested!

An [End User] is used in the CNCF sense to refer to consumers of the Kubernetes artifacts built and published by [Release Managers]

[end user]: https://www.cncf.io/enduser/

```
Scenario: End User lists the available packages from the stable channel
  Given an End User has configured the officially documented package mirror for stable releases on their machine
  When they use their system's package manager to query the list of kubelet packages available (e.g. apt-cache policy kubelet)
  Then they see a list of all stable patch versions of the kubelet and a preference to install the latest available patch version
  But they do not see any alpha, beta, release candidate or nightly releases of the kubelet package from this specific kubernetes minor version
  And they do not see any packages of any other kubernetes minor releases
```

```
Scenario: Release tools automatically publish new packages
  Given a Release Manager ran `./krel stage --nomock` and `./krel release --nomock` to create a new  release (alpha, beta, rc or final).
  When an End User inspects the officially documented deb and rpm kubernetes repositories for this specific kubernetes minor version

>   Then they see the newly cut pre releases published in the dev channel only

Reviewer 2: ```suggestion
  Then they see the newly cut releases published in the appropriate channel only
```

## Summary

> This document describes how deb and rpm packages get published to the [officially documented package mirror] as part of cutting a release, with the tooling the release management team has at hand ([krel] and others from [k/release], [k/k/build] and potentially other places).

Author: I like the idea of this KEP calling out and naming known locations where End Users can retrieve the published packages and given the proposal to have **stable**, **dev**, and **nightly bulld** channels I think this is important. 

It also prompts us to think about End User documentation and how we communicate this out.
 
I agree that 'officially' is extraneous! If it's documented, _it is official_

### Goals

The whole process should be folded into the release tooling, it should be part of the release process, and should not involve anyone other than the release managers.
For each release the release team cuts, packages should also be generated and published automatically.

There should be multiple channels users can subscribe to: **stable**, **dev**, and **nightly**.

### Non-Goals

The original version of this KEP had the following statement:

> The actual package generation is a different problem that is discussed in [this KEP][pkg-gen-kep].

Based on recent review of this KEP @saschagrunert wants SIG Release to own
the process of automatically building and publishing the packages and this will

> implicitly aceive the main goal of [#2503](https://github.com/kubernetes/enhancements/issues/2503).

Reviewer 2: Is there a typo in `aceive`, should it be `achieve`?

[pkg-gen-kep]: https://github.com/kubernetes/enhancements/pull/858

## Proposal

- Make the infrastructure generic and simple enough to be easily handed off to the CNCF

>   - Storage buckets (to store the staged/released packages)

Reviewer: Good, depending on the technical solution I'll make that statement wider and add "…or anything similar".

  - DNS entries (e.g. apt.kubernetes.io, ...)
  - package mirror (e.g. a self hosted aptly/artifactory/... or as a service)
    - have multiple channels, e.g. `stable`, `dev`, `nightly`
- Run the package builds as part of [krel stage and release]

> - Have a safe way to store the signing key and make it available to the release team and release tooling

Reviewer 3: Can we log a concern around this (maybe after merge)? It's not really fleshed out enough.

Reviewer: Yeah I think it really depends on the technical solution we choose.

- Automatically sign the repository and packages
- Automatically build and publish packages on a nightly basis

[krel stage and release]: https://github.com/kubernetes/release/blob/master/docs/krel/README.md#usage

### User Stories

_Note_: The following user stories are using the keywords of the [gherkin language][gherkin].

[gherkin]: https://cucumber.io/docs/gherkin/reference/

#### User Roles

[Release Managers] are SIG Release Team Members that are

> ... Kubernetes contributors responsible for maintaining release branches and creating releases by using the tools SIG Release provides.

[release managers]: https://kubernetes.io/releases/release-managers

An [End User] is used in the CNCF sense to refer to consumers of the Kubernetes artefacts built and published by [Release Managers]

[end user]: https://www.cncf.io/enduser/

```
Scenario: End User lists the available packages from the stable channel
  Given an End User has configured the officially documented package mirror for stable releases on their machine
  When they use their system's package manager to query the list of kubelet packages available (e.g. apt-cache policy kubelet)

>   Then they see a list of all stable patch versions of the kubelet that stem from the minor version and a preference to install the latest patch version of the kubelet

Reviewer 4: line 118: Then they see a list of all stable patch versions of the kubelet that `stem` from the

>   Then they see a list of all stable patch versions of the kubelet that stem from the minor version and a preference to install the latest patch version of the kubelet

Reviewer: Improved that line

  But they do not see any alpha, beta, release candidate or nightly releases of the kubelet package from this specific kubernetes minor version
  And they do not see any packages of any other kubernetes minor releases
```

```
Scenario: Release tools automatically publish new packages
  Given a Release Manager ran `./krel stage --nomock` and `./krel release --nomock` to create a new pre (alpha, beta, rc) release.
  When an End User inspects the officially documented deb and rpm kubernetes repositories for this specific kubernetes minor version
  Then they see the newly cut pre releases published in the dev channel only

> ```

Reviewer 2: Perhaps this story should also include stable releases, not only prereleases?

Reviewer: Agreed, fixed as suggested.

```

> Scenario: End users can retrieve the public key the packages or the repository metadata is signed with

Reviewer 3: ```suggestion
Scenario: End users can retrieve the public key that signed the packages, or the repository metadata is signed with
```

  Given an End User configured their system to disallow the use of unsigned, untrusted package repositories
  And they have setup the officially documented repository for a specific kubernetes minor release
  When they download the public key from the location stated in the official documentation
  And they configure their system's package manager to use that key
  And they use their system's package manager to install a package from this specific kubernetes minor release

>   Then their package manager will not complain about untrusted packages, sources or repositories

Reviewer 3: ```suggestion
  Then their package manager does not complain about untrusted packages, sources or repositories
  And the package install can succeed
```

        - have multiple channels, e.g. `stable`, `dev`, `nightly`
- Run the package builds as part of [krel stage and release]
- Have a safe way to store the signing key and make it available to the release team and release tooling
- Automatically sign the repository and packages
- Automatically build and publish packages on a nightly basis

[krel stage and release]: https://github.com/kubernetes/release/blob/master/docs/krel/README.md#usage

### User Stories
*Note*: The following user stories are using the keywords of the [gherkin language][gherkin].

[gherkin]: https://cucumber.io/docs/gherkin/reference/

#### User Roles
[Release Managers] are SIG Release Team Members that are 
>  ... Kubernetes contributors responsible for maintaining release branches and creating releases by using the tools SIG Release provides.

[Release Managers]: https://kubernetes.io/releases/release-managers

An [End User] is used in the CNCF sense to refer to consumers of the Kubernetes artefacts built and published by [Release Managers]
[End User]: https://www.cncf.io/enduser/

```
Scenario: End User lists the available packages from the stable channel

>   Given an End User has configured the officially documented package mirror for stable releases for a specific kubernetes minor version ("the minor version") on their machine

Reviewer: Fixed, removed the above quoted part. I agree we should not distinct between distributions for now.

  When they use their system's package manager to query the list of kubelet packages available (e.g. apt-cache policy kubelet)
  Then they see a list of all stable patch versions of the kubelet that stem from the minor version and a preference to install the latest patch version of the kubelet
  But they do not see any alpha, beta, release candidate or nightly releases of the kubelet package from this specific kubernetes minor version
  And they do not see any packages of any other kubernetes minor releases
```

```
Scenario: Release tools automatically publish new packages
  Given a Release Manager ran `./gcbmgr stage master --build-at-head --nomock`
  And a Release Branch Manager ran `./gcbmgr release master --buildversion=${{VERSIONID}} --nomock`
  When an End User inspects the officially documented deb and rpm kubernetes repositories for this specific kubernetes minor version

>   Then they see the newly cut alpha releases published in the alpha channel only

Reviewer: Fixed :+1:

> 

Author: @saschagrunert @puerco what are your thoughts @detiber's comment here ^^ Based on a quick reading of https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/kubeadm/2503-Artifact-Generation there are some good User Stories there that I think we could either incorporate (or reject) also there is a a reference to some tooling [fpm ](https://github.com/jordansissel/fpm)that could be reviewed for use here?

> 

Reviewer: I think it's a good idea to use KEP 2503 as inspiration if it does not unnecessarily broaden the scope of this enhancement.

> 

Reviewer: I'll say we leave the implementation details outside of this KEP and state the overall goal.

### User Stories
*Note*: The following user stories are using the keywords of the [gherkin language][gherkin].

[gherkin]: https://cucumber.io/docs/gherkin/reference/

#### User Roles
[Release Managers] are SIG Release Team Members that are

> > ... responsible for maintaining release branches, tagging releases, and building/packaging Kubernetes.

Author: +1 on being precise. This is quoted text taken from here ...

https://github.com/kubernetes/website/edit/main/content/en/releases/release-managers.md

But I thought that originally I was quoting from a sig-release file describing the Release Managers and am struggling to find that now. 

I am happy to update the source doc(s)

> > ... responsible for maintaining release branches, tagging releases, and building/packaging Kubernetes.

Reviewer: I'll provide an update to this page in a bit, good catch!

[Release Managers]: https://github.com/kubernetes/website/blob/main/content/en/releases/release-managers.md

An [End User] is used in the CNCF sense to refer to consumers of the Kubernetes artefacts built and published by [Release Managers]
[End User]: https://www.cncf.io/enduser/

```

> Scenario: End User lists the available kubelets from the stable channel

Reviewer 5: ```suggestion
Scenario: End User lists the available packages from the stable channel
```

Author: Nice catch!

### Implementation Details/Notes/Constraints

Packages will be published for different:

- Managers, like `apt` (consumes deb packages) and `yum` (consumes rpm packages).
- **`${k8s_release}`**: the version of kubernetes `<major>.<minor>`
  (e.g. `1.12`, `1.13`, `1.14`, ...)
- **`${channel}`**: can be `stable`, `dev`, `nightly`
  - `stable`: all official releases for `${k8s_release}`
    (e.g.: `1.13.0`, `1.13.1`, `1.13.2`, ...)
  - `dev`: all development releases for all minor releases in this `${k8s_release}`, including `alpha`s, `beta`s and `rc`s
    (e.g.: `1.13.0-rc.2`, `1.13.2-beta.0`, `1.13.1-alpha.3`, ...)
  - `nightly`: any package cut automatically on a daily basis (optionally)

This means, that End Users can configure their systems’ package managers to use
those different `${channel}`s of a kubernetes `${k8s_release}` for their
corresponding package manager.

A configuration for the package managers might look something like:

- deb:
  ```
  # deb http://apt.kubernetes.io ${k8s_release} ${channel}

>   deb http://apt.kubernetes.io/debian-jessie 1.13 nightly

Reviewer 3: Aside: I'd expect to see something like
```
deb [signed-by=/etc/keyrings/kubernetes-keyring.gpg] http://apt.kubernetes.io/debian-jessie 1.26 nightly
```

Reviewer: Good, I'll add this example by modifying the existing one.

  ```
- rpm/yum:
  ```
  [kubernetes]
  name=Kubernetes
  # baseurl=http://yum.kubernetes.io/${k8s_release}/${channel}
  baseurl=http://yum.kubernetes.io/fedora-27/1.13/nightly
  enabled=1
  gpgcheck=1
  repo_gpgcheck=1
  gpgkey=file:///etc/pki/rpm-gpg/kubernetes.gpg.pub
  ```

Different architectures will be published into the same repos, it is up to the package managers to pull and install the correct package for the target platform.

Ideally and optionally, publishing/promoting a package means to commit a change
to a configuration file which triggers a "package promotion tool", which:

- manages which packages need to go into which `${channel}` for which package manager of which `${k8s_release}`

> - guard that by the packages checksum

Reviewer: I'd like to leave that out for now since it moves the KEP more into the SLSA direction (where we already have https://github.com/kubernetes/enhancements/issues/3027 for).

### Implementation Details/Notes/Constraints

Packages will be published for different:

- Package managers, like `apt` (consumes deb packages) and `yum` (consumes rpm packages).
- **`${k8s_release}`**: the version of kubernetes `<major>.<minor>`
  (e.g. `1.12`, `1.13`, `1.14`, ...)
- **`${channel}`**: can be `stable`, `dev`, `nightly`
  - `stable`: all official releases for `${k8s_release}`
    (e.g.: `1.13.0`, `1.13.1`, `1.13.2`, ...)
  - `dev`: all development releases for all minor releases in this `${k8s_release}`, including `alpha`s, `beta`s and `rc`s
    (e.g.: `1.13.0-rc.2`, `1.13.2-beta.0`, `1.13.1-alpha.3`, ...)
  - `nightly`: any package cut automatically on a daily basis (optionally)

This means, that End Users can configure their systems’ package managers to use
those different `${channel}`s of a kubernetes `${k8s_release}` for their
corresponding package manager.

A configuration for the package managers might look something like:

- deb:
  ```
  # deb http://apt.kubernetes.io ${k8s_release} ${channel}

>   deb [signed-by=/etc/keyrings/kubernetes-keyring.gpg] http://apt.kubernetes.io/debian-jessie 1.26 nightly

Reviewer 2: AFAIK, we will avoid having a repo per distro version.

```suggestion
  deb [signed-by=/etc/keyrings/kubernetes-keyring.gpg] http://apt.kubernetes.io/debian 1.26 nightly
```

  ```
- rpm/yum:
  ```
  [kubernetes]
  name=Kubernetes
  # baseurl=http://yum.kubernetes.io/${k8s_release}/${channel}

>   baseurl=http://yum.kubernetes.io/fedora-27/1.26/nightly

Reviewer 2: Same as above.

```suggestion
  baseurl=http://yum.kubernetes.io/fedora/1.26/nightly
```

>     - `nightly`: any package cut automatically on a daily basis

Reviewer: Fixed, we should leave dist out for now.

Different architectures will be published into the same repos, it is up to the package managers to pull and install the correct package for the target platform.

Ideally, publishing/promoting a package means to commit a change to a configuration file which triggers a "package promotion tool".
That tool ...

> - manages which packages need to go into which `${channel}` for which `${dist}` of which `${k8s_release}`

Author: Ok. Need some clarification here, when we talk about "file promotion" what do we mean? Is that promotion of packages through the channels from nightly, to dev to stable channels? Or are we talking about the package contents so that end users can see what files have changed in a package?

> - manages which packages need to go into which `${channel}` for which `${dist}` of which `${k8s_release}`

Reviewer: With file promotion I'm referring to this effort: https://github.com/kubernetes-sigs/promo-tools/issues/413 

We do not use this process yet but plan to for the binary artifacts later on. I don't think we have to wire this effort into the deb/rpm story yet to keep things separated.

> - manages which packages need to go into which `${channel}` for which `${dist}` of which `${k8s_release}`

Reviewer 6: OK, I disagree with both using and not using file promotion. Publishing the packages is a different beast altogether, ie we cannot use a simple copy or file promotion. 

There is a process that needs to run after forging a new set of packages both in .rpm (`updaterepo`) and .deb (`dpkg-scanpackages`). the new packages need to land in their repositories and then the respective tool needs to be run to update the repo's index.

If we do this as part of a promotion process (and I could do a strong argument about favoring this design), we could add the index update to kpromo (the "image promoter").

On the other hand we cannot "do nothing". The process needs to run to update the indices to that OS tools (apt/yum/dnf) can pick up updates.

I don't think the KEP would be complete without defining how we plan to do this. The end-user story is not complete without finishing this part.

> - manages which packages need to go into which `${channel}` for which `${dist}` of which `${k8s_release}`

Reviewer 6: @detiber Re:

> If we end up using one or more third party provider(s) for hosting deb/rpm packages

I think we should still dump ideas on the enhancement discussion to plan on building our own repo. I'm not against using a third party by any means, but I think if we choose to go that route we will need to write another KEP. This one is specific to publishing on community infra. WDYT?

> - manages which packages need to go into which `${channel}` for which `${dist}` of which `${k8s_release}`

Author: cc @justaugustus

> - manages which packages need to go into which `${channel}` for which `${dist}` of which `${k8s_release}`

Reviewer 7: @puerco my though was before we go down the path of spinning up community run infrastructure and taking on the complications related to that, especially around gpg key management, we could do an initial spike using third party hosting first, especially since it could potentially save us a lot of implementation, on going maintenance, and security related toil.

That way, in the event that third party hosting does not work for us, we could then start getting into the detailed implementation details around what is needed, and we could update the KEP with the more detailed plan.

> - manages which packages need to go into which `${channel}` for which `${dist}` of which `${k8s_release}`

Reviewer: I'll leave most of the details out for now and make the promotion optional (since it's based on a different story).

### Risks and Mitigations

- _Risk_: We don't find a proper way to share secrets like the signing key
  _Mitigation_: Using a third party tool like 1Password
- _Risk_: Building all the packages for all the distributions and their version takes too long to be done nightly or via cutting the release  
  _Mitigation_: We do not deliver nightly packages.

## Design Details

### Test Plan

There should be post-publish tests, which can be run as part or after the release process

- pull packages from the official mirrors
- assert that all the packages we expect to be published are actually published

> - assert that all publsihed packages have the expected checksums

Reviewer 3: ```suggestion
- assert that all published packages have the expected checksums
```

- assert that the packages and the repo metadata is signed with the current signing key
- assert that installed packages contain the correct binary versions

### Graduation Criteria

> In general we can keep the current way of publishing (via googlers onto Google's infrastructure) and introduce new infrastructure in parallel.

Reviewer 8: IMHO: we should keep *publishing* but builds should be shifted to community control instead of Googler workstations. If that's the intention, I don't see it explicitly called out.

The existing signing+publishing can easily be switched to consume community built packages, they're distinct steps / tools in the scripts.

Reviewer: This depends on which technical solution we choose. I added this paragraph to make it more clear that we can iteratively move forward: https://github.com/kubernetes/enhancements/pull/3434/files#diff-9356b6835a7626094e8636d5f5a8a1d71488e4d2abb948557ce41df3b77f88a1R204-R207

I think in the end it should be overall community owned, which would leave the current statement untouched, right?

> In general we can keep the current way of publishing (via googlers onto Google's infrastructure) and introduce new infrastructure in parallel.

Reviewer 8: https://github.com/kubernetes/release/issues/913#issuecomment-1276364097 is also worth considering as an immediate stopgap, KEP or otherwise.

Reviewer: I added one additional commit on top of the PR to explicitly mention the parallel improvement path. I'll make another comment about https://github.com/kubernetes/release/issues/913#issuecomment-1276364097 on the corresponding issue.

> 

Reviewer: I added possible mitigations to the steps as suggested.

### Test Plan

There should be post-publish tests, which can be run as part or after the release process

> - pull packages from the official mirrors (via the [redirector] if in place)

Reviewer: Fixed

- assert that all the packages we expect to be published are actually published
- assert that all publsihed packages have the expected checksums 

> - assert that the packages and the repo metadata is signed with the current signing key

Reviewer: Yep, added.

> 

Reviewer 7: I think it would be good to update alpha requirements to include:
- There is a documented process to create and publish deb and rpm packages of Kubernetes components
- It is possible to consume the published deb and rpm packages using steps similar to the documented process

Author: Yes to this! Why 'similar' though, should the docs not offer a definitive guide?

Reviewer 7: > Yes to this! Why 'similar' though, should the docs not offer a definitive guide?

My thought was that for Alpha we would not update the docs to reference the new package repositories. In which case some modifications would be needed for testing to work around the docs referencing old gpg keys, repositories, etc.

#### Alpha -> Beta Graduation

> - [ ] [krel] creates and publishes deb and rpm packages of Kubernetes components

Reviewer 6: Here, again, I think we need to add two more items:

```suggestion
- [ ] [krel] creates and deb and rpm packages of Kubernetes components
- [ ] Packages are signed
- [ ] Repository indices are updated after packages are copied to the repositories
```

I am intentionally:
a) Splitting signing to a separate item
b) Not adding implementation details of how we sign or how we copy+update

Author: Great feedback,  TYSM @puerco !

- [ ] post-publish tests are written and run as part of the release process
- [ ] nightly builds will be built and published on a daily basis using [krel] which will be improved to take over this task from [kubepkg] making use of [pre-existing periodic jobs] (https://github.com/kubernetes/test-infra/blob/97cb34fa9e2bfc4af35de3e561cb9fc5a1094da1/config/jobs/kubernetes/sig-release/kubernetes-builds.yaml#L120-L166)
- [ ] documentation written checked to be complete and correct. 

[kubepgk]: https://github.com/kubernetes/release/tree/master/cmd/kubepkg 

> [krel]: https://github.com/kubernetes/release/blob/master/docs/krel/README.md 

Reviewer 7: I think Beta graduation should likely include publicly publishing documentation updates to offer an alternative option to consume the community owned packages, and publishing packages using both the old and new processes.

This would provide us the opportunity to get end-user driven feedback prior to GA

Author: Ah OK, I did not know that this would bring about an End User process change resulting in a new procedure for packages published using this implementation I had assumed no process change for End Users @kubernetes/sig-release 

If there is a process change that should be called out for sure.

We already have in Beta -> GA Graduation 
>  documentation written checked to be complete and correct.

#### Beta -> GA Graduation

> 

Reviewer 7: GA Graduation should probably require:
- Removing remaining documentation for the old google hosted packages

Reviewer 6: Plus deprecation policy + comms plan

Author: I've added headers for Documentation removal, Deprecation policy, and Comns plans.

Reviewer 2: > Plus deprecation policy + comms plan

I think we need to do this much earlier. We should probably have some initial deprecation policy ideas outlined in this KEP. For comms, I think we should start communicating as early as possible, probably when graduating to beta. For example:

* Announce that we started efforts on publishing packages to community infra, provide semi-frequent updates to the commucnity
* Once we graduate to beta, start promoting new infrastructure via all communication channels. Ask users to give it a try and provide us feedback
* When graduating to GA, start the deprecation period and communicate that via all communication channels, including CNCF

Additionally, I think we might need some sort of survey in beta or when graduating to GA. We should collect information from users did they tested new infra, how it works for them, and how much effort is it to migrate to new infra for them.

#### Alpha -> Beta Graduation

- [ ] [kubepkg] [creates deb and rpm packages of Kubernetes componets]
- [ ] [pkg-gen-kep] and published those packages as part of the release process
- [ ] post-publish tests are written and run as part of the release process

> - [ ] nightly builds will be built and published on a daily basis

Reviewer 5: @saschagrunert maybe ` nightly builds will be built and published on a daily basis via [krel]` ? since `kubepkg` will be merged in `krel`.

> - [ ] nightly builds will be built and published on a daily basis

Reviewer: @ameukam I'm fine with that. I was just thinking that we already have a kubepkg periodics which may be reused: 

https://github.com/kubernetes/test-infra/blob/97cb34fa9e2bfc4af35de3e561cb9fc5a1094da1/config/jobs/kubernetes/sig-release/kubernetes-builds.yaml#L120-L166

Naming them towards `krel` makes sense to me :+1:

#### Alpha -> Beta Graduation

- [ ] [kubepkg] [creates deb and rpm packages of Kubernetes componets]
- [ ] [pkg-gen-kep] and published those packages as part of the release process
- [ ] post-publish tests are written and run as part of the release process
- [ ] nightly builds will be built and published on a daily basis
- [ ] documentation written checked to be complete and correct. 

> [kubepgk]: https://github.com/kubernetes/release/tree/master/cmd/kubepkg 

Reviewer 5: ```suggestion
[kubepkg]: https://github.com/kubernetes/release/tree/master/cmd/kubepkg 
```
